services/kvstore/src/main/java/tech/pegasys/pantheon/services/kvstore/InMemoryKeyValueStorage.java

```diff
@@ -39,11 +41,43 @@
     }
   }
 
+  @Override
+  public boolean containsKey(final BytesValue key) throws StorageException {
+    final Lock lock = rwLock.readLock();
+    lock.lock();
+    try {
+      return hashValueStore.containsKey(key);
+    } finally {
+      lock.unlock();
+    }
+  }
+
   @Override
   public Transaction startTransaction() {
     return new InMemoryTransaction();
   }
 
+  @Override
+  public long removeUnless(final Predicate<BytesValue> inUseCheck) {
+    hashValueStore.keySet().removeIf(key -> !inUseCheck.test(key));
+    return 0;
+  }
+
+  @Override
+  public void clear() {
+    final Lock lock = rwLock.writeLock();
+    lock.lock();
+    try {
+      hashValueStore.clear();
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  public Set<BytesValue> keySet() {
```

Comment:
Do we want a snapshot of the keyset or one that morphs as the underlying datastore morphs?  That would invalidate iteration too if the state changes.

Reply:
It would be consistent with the other kvstores to have snapshot iteration, which I believe this is. Any motivating examples for the alternative?

Reply:
If we want a point in time we will need to copy what is returned from `keySet` instead of just wrapping it in unmodifiableSet.  The wrapper only prevents methods from adding/removing, not from the set being wrapped changing, which is what we would see here.